developer/src/kmc-keyboard/src/compiler/layr.ts

```diff
@@ -18,28 +17,62 @@ export class LayrCompiler extends SectionCompiler {
 
   public validate() {
     let valid = true;
-    if(!this.keyboard.layers?.[0]?.layer?.length) {
+    if (!this.keyboard.layers?.[0]?.layer?.length) {
       valid = false;
       this.callbacks.reportMessage(CompilerMessages.Error_MustBeAtLeastOneLayerElement());
     }
-    // TODO-LDML
+    let hardwareLayers = 0;
+    this.keyboard.layers.forEach(({ hardware, form }) => {
+      // TODO-LDML: in the future >1 hardware layer may be allowed, check for duplicates
+      if (form === 'touch') {
+        if (hardware) {
+          valid = false;
+          this.callbacks.reportMessage(CompilerMessages.Error_InvalidFile({
+            errorText: `Not allowed: hardware="${hardware}" with layers form="touch"`
+          }));
+        }
+      } else if (form === 'hardware') {
+        hardwareLayers++;
+        if (!hardware) {
+          valid = false;
+          this.callbacks.reportMessage(CompilerMessages.Error_InvalidFile({
+            errorText: `on layers form="hardware", missing required hardware= attribute.`
+          }));
+        } else if (!constants.layr_list_hardware_map.get(hardware)) {
+          valid = false;
+          this.callbacks.reportMessage(CompilerMessages.Error_InvalidFile({
+            errorText: `Unknown hardware layout id: hardware="${hardware}"`
+          }));
+        } else if (hardwareLayers > 1) { // TODO-LDML: revisit if spec changes
+          valid = false;
+          this.callbacks.reportMessage(CompilerMessages.Error_MustHaveAtMostOneLayersElementPerForm({ form }));
+        }
+      } else {
+        // Should not be reached due to XML validation.
+        valid = false;
+        this.callbacks.reportMessage(CompilerMessages.Error_InvalidFile({
+          errorText: `Invalid form="${form}" on layers element`
```

Comment:
Ditto, separate message. It feels like busy work but helps in the long run because each error is uniquely identified and much easier to track.

Reply:
This specific one is almost an internal error, since the XML validation catches it:

```
ERROR 300007: The source file has an invalid structure: Error: /keyboard/layers/0/form: enum: must be equal to one of the allowed values {"allowedValues":["hardware","touch"]}
```

Reply:
I've noted as an internal error. LMK if you want to revisit.